File: gear-commons/src/main/scala/cn/howardliu/gear/commons/utils/Processor.scala
package cn.howardliu.gear.commons.utils

import java.io.{InputStreamReader, InputStream}
import java.nio.charset.Charset

import org.apache.commons.io.IOUtils

/**
 * <br>created at 16-12-27
 * @author liuxh
 * @since 1.0.1
 */
object Processor {
  def execute(cmd: String): String = {
    var in: InputStream = null
    var process: Process = null
    try {
      process = Runtime.getRuntime.exec(cmd)
      in = process.getInputStream
      IOUtils.toString(new InputStreamReader(in, Charset.defaultCharset))
    } catch {
      case ex: Exception => "(error executing: " + cmd + ")"
      case err: Error =>
        if (err.getMessage != null && (err.getMessage.contains("posix_spawn") || err.getMessage.contains("UNIXProcess"))) {
          return "(error executing: " + cmd + ")"
        }
        throw err
    } finally {
      if (process != null) {
        IOUtils.closeQuietly(process.getOutputStream)
        IOUtils.closeQuietly(process.getInputStream)
        IOUtils.closeQuietly(process.getErrorStream)
      }
    }
  }

  def execute(cmd: Array[String]): String = {
    var in: InputStream = null
    var process: Process = null
    try {
      process = Runtime.getRuntime.exec(cmd)
      in = process.getInputStream
      IOUtils.toString(new InputStreamReader(in, Charset.defaultCharset))
    } catch {
      case ex: Exception => "(error executing: " + cmd + ")"
      case err: Error =>
        if (err.getMessage != null && (err.getMessage.contains("posix_spawn") || err.getMessage.contains("UNIXProcess"))) {
          return "(error executing: " + cmd + ")"
        }
        throw err
    } finally {
      if (process != null) {
        IOUtils.closeQuietly(process.getOutputStream)
        IOUtils.closeQuietly(process.getInputStream)
        IOUtils.closeQuietly(process.getErrorStream)
      }
    }
  }
}

File: gear-commons/src/main/java/cn/howardliu/gear/commons/utils/Processor.java
package cn.howardliu.gear.commons.utils;

import org.apache.commons.io.IOUtils;

import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.Charset;
import java.util.Arrays;

/**
 * <br>created at 2020/5/5
 *
 * @author liuxh
 * @since 1.0.0
 */
public class Processor {
    public static String execute(String cmd) {
        Process process = null;
        try {
            process = Runtime.getRuntime().exec(cmd);
            InputStream in = process.getInputStream();
            return IOUtils.toString(new InputStreamReader(in, Charset.defaultCharset()));
        } catch (Exception e) {
            return "(error executing: " + cmd + ")";
        } catch (Error e) {
            final String message = e.getMessage();
            if (message != null && (message.contains("posix_spawn") || message.contains("UNIXProcess"))) {
                return "(error executing: " + cmd + ")";
            }
            throw e;
        } finally {
            if (process != null) {
                IOUtils.closeQuietly(process.getOutputStream());
                IOUtils.closeQuietly(process.getInputStream());
                IOUtils.closeQuietly(process.getErrorStream());
            }
        }
    }

    public static String execute(String[] cmd) {
        final String cmdArrayStr = Arrays.toString(cmd);
        Process process = null;
        try {
            process = Runtime.getRuntime().exec(cmd);
            InputStream in = process.getInputStream();
            return IOUtils.toString(new InputStreamReader(in, Charset.defaultCharset()));
        } catch (Exception e) {
            return "(error executing: " + cmdArrayStr + ")";
        } catch (Error e) {
            final String message = e.getMessage();
            if (message != null && (message.contains("posix_spawn") || message.contains("UNIXProcess"))) {
                return "(error executing: " + cmdArrayStr + ")";
            }
            throw e;
        } finally {
            if (process != null) {
                IOUtils.closeQuietly(process.getOutputStream());
                IOUtils.closeQuietly(process.getInputStream());
                IOUtils.closeQuietly(process.getErrorStream());
            }
        }
    }
}
